File: sendfn/typescript/src/events/aws-sns-verifier.ts
import { createVerify } from 'node:crypto';
import { SendfnError } from '../errors';

export interface SnsMessage {
  Type?: string;
  MessageId?: string;
  TopicArn?: string;
  Subject?: string;
  Message?: string;
  Timestamp?: string;
  SignatureVersion?: string;
  Signature?: string;
  SigningCertURL?: string;
  UnsubscribeURL?: string;
  Token?: string;
  SubscribeURL?: string;
}

export interface AwsSnsVerifierOptions {
  now?: () => Date;
  fetchCertificate?: (url: string) => Promise<string>;
  verifySignature?: (canonicalMessage: string, signature: string, certificate: string) => Promise<boolean> | boolean;
  maxAgeMs?: number;
  topicArns?: readonly string[];
  confirmSubscription?: (url: string) => Promise<void>;
}

const CERT_HOST_PATTERN = /^sns\.[a-z0-9-]+\.amazonaws\.com$/i;

function createWebhookError(code: string, message: string): SendfnError {
  return new SendfnError(message, {
    code,
    retryable: false,
  });
}

export class AwsSnsVerifier {
  private readonly now: () => Date;
  private readonly fetchCertificate: (url: string) => Promise<string>;
  private readonly verifySignature: (canonicalMessage: string, signature: string, certificate: string) => Promise<boolean> | boolean;
  private readonly maxAgeMs?: number;
  private readonly topicArns: Set<string>;
  private readonly confirm: (url: string) => Promise<void>;

  constructor(options: AwsSnsVerifierOptions = {}) {
    this.now = options.now ?? (() => new Date());
    this.fetchCertificate = options.fetchCertificate ?? this.defaultFetchCertificate;
    this.verifySignature = options.verifySignature ?? this.defaultVerifySignature;
    this.maxAgeMs = options.maxAgeMs;
    this.topicArns = new Set(options.topicArns ?? []);
    this.confirm = options.confirmSubscription ?? this.defaultConfirmSubscription;
  }

  async confirmSubscription(message: SnsMessage): Promise<void> {
    if (message.Type !== 'SubscriptionConfirmation' || !message.SubscribeURL) {
      throw createWebhookError('SENDFN_WEBHOOK_MESSAGE_INVALID', 'SNS message is malformed');
    }
    this.validateSnsUrl(message.SubscribeURL);
    await this.confirm(message.SubscribeURL);
  }

  async verify(message: SnsMessage): Promise<void> {
    this.validateEnvelopeShape(message);
    if (!this.topicArns.has(message.TopicArn!)) {
      throw createWebhookError('SENDFN_WEBHOOK_MESSAGE_INVALID', 'SNS message is malformed');
    }
    this.validateSigningCertUrl(message.SigningCertURL!);
    this.validateTimestamp(message.Timestamp!);

    const canonicalMessage = this.buildCanonicalMessage(message);
    const certificate = await this.fetchCertificate(message.SigningCertURL!);

    let isValid = false;
    try {
      isValid = await this.verifySignature(canonicalMessage, message.Signature!, certificate);
    } catch {
      throw createWebhookError('SENDFN_WEBHOOK_SIGNATURE_INVALID', 'SNS signature verification failed');
    }

    if (!isValid) {
      throw createWebhookError('SENDFN_WEBHOOK_SIGNATURE_INVALID', 'SNS signature verification failed');
    }
  }

  private validateEnvelopeShape(message: SnsMessage): void {
    const requiredFields: Array<keyof SnsMessage> = [
      'Type',
      'Message',
      'MessageId',
      'TopicArn',
      'Timestamp',
      'SignatureVersion',
      'Signature',
      'SigningCertURL',
    ];

    for (const field of requiredFields) {
      if (typeof message[field] !== 'string' || message[field]!.trim() === '') {
        throw createWebhookError('SENDFN_WEBHOOK_MESSAGE_INVALID', 'SNS message is malformed');
      }
    }

    if (message.Type !== 'Notification' && message.Type !== 'SubscriptionConfirmation') {
      throw createWebhookError('SENDFN_WEBHOOK_MESSAGE_INVALID', 'SNS message is malformed');
    }

    if (message.Type === 'SubscriptionConfirmation') {
      for (const field of ['Token', 'SubscribeURL'] as const) {
        if (typeof message[field] !== 'string' || message[field]!.trim() === '') {
          throw createWebhookError('SENDFN_WEBHOOK_MESSAGE_INVALID', 'SNS message is malformed');
        }
      }
      this.validateSnsUrl(message.SubscribeURL!);
    }

    if (message.SignatureVersion !== '1') {
      throw createWebhookError('SENDFN_WEBHOOK_MESSAGE_INVALID', 'SNS message is malformed');
    }
  }

  private validateSigningCertUrl(value: string): void {
    let url: URL;
    try {
      url = new URL(value);
    } catch {
      throw createWebhookError('SENDFN_WEBHOOK_SIGNATURE_INVALID', 'SNS signature verification failed');
    }

    if (url.protocol !== 'https:' || !CERT_HOST_PATTERN.test(url.hostname)) {
      throw createWebhookError('SENDFN_WEBHOOK_SIGNATURE_INVALID', 'SNS signature verification failed');
    }
  }

  private validateTimestamp(timestamp: string): void {
    const parsed = new Date(timestamp);
    if (Number.isNaN(parsed.getTime())) {
      throw createWebhookError('SENDFN_WEBHOOK_MESSAGE_INVALID', 'SNS message is malformed');
    }

    if (this.maxAgeMs !== undefined && Math.abs(this.now().getTime() - parsed.getTime()) > this.maxAgeMs) {
      throw createWebhookError('SENDFN_WEBHOOK_MESSAGE_INVALID', 'SNS message is malformed');
    }
  }

  private buildCanonicalMessage(message: SnsMessage): string {
    const orderedFields: Array<keyof SnsMessage> = message.Type === 'SubscriptionConfirmation'
      ? ['Message', 'MessageId', 'SubscribeURL', 'Timestamp', 'Token', 'TopicArn', 'Type']
      : ['Message', 'MessageId', ...(message.Subject ? ['Subject' as const] : []), 'Timestamp', 'TopicArn', 'Type'];

    return `${orderedFields
      .map((field) => `${field}\n${message[field] ?? ''}`)
      .join('\n')}\n`;
  }

  private validateSnsUrl(value: string): void {
    let url: URL;
    try {
      url = new URL(value);
    } catch {
      throw createWebhookError('SENDFN_WEBHOOK_MESSAGE_INVALID', 'SNS message is malformed');
    }
    if (url.protocol !== 'https:' || !CERT_HOST_PATTERN.test(url.hostname)) {
      throw createWebhookError('SENDFN_WEBHOOK_MESSAGE_INVALID', 'SNS message is malformed');
    }
  }

  private async defaultConfirmSubscription(url: string): Promise<void> {
    const response = await fetch(url, { redirect: 'error' });
    if (!response.ok) {
      throw createWebhookError('SENDFN_WEBHOOK_CONFIRMATION_FAILED', 'SNS subscription confirmation failed');
    }
  }

  private async defaultFetchCertificate(url: string): Promise<string> {
    const response = await fetch(url);
    if (!response.ok) {
      throw createWebhookError('SENDFN_WEBHOOK_SIGNATURE_INVALID', 'SNS signature verification failed');
    }

    return response.text();
  }

  private defaultVerifySignature(canonicalMessage: string, signature: string, certificate: string): boolean {
    const verifier = createVerify('RSA-SHA1');
    verifier.update(canonicalMessage, 'utf8');
    verifier.end();
    return verifier.verify(certificate, Buffer.from(signature, 'base64'));
  }
}

File: sendfn/python/sendfn/events/aws_sns_verifier.py
"""AWS SNS signature verification helpers for SES webhooks."""

from __future__ import annotations

import asyncio
import base64
import re
from datetime import datetime, timezone
from typing import Any, Callable, Iterable, Optional
from urllib.parse import urlparse
from urllib.request import urlopen

from cryptography import x509
from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import padding, rsa

from ..errors import SendfnError

CERT_HOST_PATTERN = re.compile(r"^sns\.[a-z0-9-]+\.amazonaws\.com$", re.IGNORECASE)


def create_webhook_error(code: str, message: str) -> SendfnError:
    """Create a stable webhook error."""
    return SendfnError(message, code=code, retryable=False)


class AwsSnsVerifier:
    """Verify AWS SNS envelopes before SES lifecycle processing."""

    def __init__(
        self,
        *,
        now: Optional[Callable[[], datetime]] = None,
        fetch_certificate: Optional[Callable[[str], str]] = None,
        verify_signature: Optional[Callable[[str, str, str], bool]] = None,
        confirm_subscription: Optional[Callable[[str], None]] = None,
        max_age_seconds: Optional[int] = None,
        topic_arns: Optional[Iterable[str]] = None,
    ) -> None:
        self.now = now or (lambda: datetime.now(timezone.utc))
        self.fetch_certificate = fetch_certificate or self._default_fetch_certificate
        self.verify_signature = verify_signature or self._default_verify_signature
        self._confirm_subscription = confirm_subscription or self._default_confirm_subscription
        self.max_age_seconds = max_age_seconds
        self.topic_arns = set(topic_arns or ())

    async def verify(self, message: dict[str, Any]) -> None:
        """Verify SNS envelope shape, freshness, cert host, and signature."""
        self._validate_envelope_shape(message)
        if message["TopicArn"] not in self.topic_arns:
            raise create_webhook_error(
                "SENDFN_WEBHOOK_MESSAGE_INVALID",
                "SNS message is malformed",
            )
        self._validate_signing_cert_url(message["SigningCertURL"])
        self._validate_timestamp(message["Timestamp"])

        canonical_message = self._build_canonical_message(message)
        certificate = await asyncio.to_thread(
            self.fetch_certificate,
            message["SigningCertURL"],
        )

        try:
            is_valid = self.verify_signature(
                canonical_message,
                message["Signature"],
                certificate,
            )
        except Exception as exc:  # pragma: no cover - fail-closed branch
            raise create_webhook_error(
                "SENDFN_WEBHOOK_SIGNATURE_INVALID",
                "SNS signature verification failed",
            ) from exc

        if not is_valid:
            raise create_webhook_error(
                "SENDFN_WEBHOOK_SIGNATURE_INVALID",
                "SNS signature verification failed",
            )

    async def confirm_subscription(self, message: dict[str, Any]) -> None:
        """Confirm a verified SNS subscription handshake."""
        if message.get("Type") != "SubscriptionConfirmation" or not message.get("SubscribeURL"):
            raise create_webhook_error(
                "SENDFN_WEBHOOK_MESSAGE_INVALID",
                "SNS message is malformed",
            )
        self._validate_sns_url(str(message["SubscribeURL"]))
        try:
            await asyncio.to_thread(
                self._confirm_subscription,
                str(message["SubscribeURL"]),
            )
        except SendfnError:
            raise
        except Exception as exc:
            raise create_webhook_error(
                "SENDFN_WEBHOOK_CONFIRMATION_FAILED",
                "SNS subscription confirmation failed",
            ) from exc

    def _validate_envelope_shape(self, message: dict[str, Any]) -> None:
        required_fields = [
            "Type",
            "Message",
            "MessageId",
            "TopicArn",
            "Timestamp",
            "SignatureVersion",
            "Signature",
            "SigningCertURL",
        ]
        for field in required_fields:
            value = message.get(field)
            if not isinstance(value, str) or not value.strip():
                raise create_webhook_error(
                    "SENDFN_WEBHOOK_MESSAGE_INVALID",
                    "SNS message is malformed",
                )

        if message["Type"] not in {"Notification", "SubscriptionConfirmation"} or message["SignatureVersion"] != "1":
            raise create_webhook_error(
                "SENDFN_WEBHOOK_MESSAGE_INVALID",
                "SNS message is malformed",
            )
        if message["Type"] == "SubscriptionConfirmation":
            for field in ("Token", "SubscribeURL"):
                value = message.get(field)
                if not isinstance(value, str) or not value.strip():
                    raise create_webhook_error(
                        "SENDFN_WEBHOOK_MESSAGE_INVALID",
                        "SNS message is malformed",
                    )
            self._validate_sns_url(message["SubscribeURL"])

    def _validate_signing_cert_url(self, signing_cert_url: str) -> None:
        parsed = urlparse(signing_cert_url)
        if parsed.scheme != "https" or not CERT_HOST_PATTERN.match(parsed.hostname or ""):
            raise create_webhook_error(
                "SENDFN_WEBHOOK_SIGNATURE_INVALID",
                "SNS signature verification failed",
            )

    def _validate_timestamp(self, timestamp: str) -> None:
        parsed = self._parse_timestamp(timestamp)
        age = abs((self.now() - parsed).total_seconds())
        if self.max_age_seconds is not None and age > self.max_age_seconds:
            raise create_webhook_error(
                "SENDFN_WEBHOOK_MESSAGE_INVALID",
                "SNS message is malformed",
            )

    def _parse_timestamp(self, timestamp: str) -> datetime:
        normalized = timestamp.replace("Z", "+00:00")
        try:
            parsed = datetime.fromisoformat(normalized)
        except ValueError as exc:
            raise create_webhook_error(
                "SENDFN_WEBHOOK_MESSAGE_INVALID",
                "SNS message is malformed",
            ) from exc

        if parsed.tzinfo is None:
            parsed = parsed.replace(tzinfo=timezone.utc)
        return parsed.astimezone(timezone.utc)

    def _build_canonical_message(self, message: dict[str, Any]) -> str:
        if message["Type"] == "SubscriptionConfirmation":
            ordered_fields = ["Message", "MessageId", "SubscribeURL", "Timestamp", "Token", "TopicArn", "Type"]
        else:
            ordered_fields = ["Message", "MessageId"]
            if message.get("Subject"):
                ordered_fields.append("Subject")
            ordered_fields.extend(["Timestamp", "TopicArn", "Type"])
        return "\n".join(f"{field}\n{message.get(field, '')}" for field in ordered_fields) + "\n"

    def _validate_sns_url(self, value: str) -> None:
        parsed = urlparse(value)
        if parsed.scheme != "https" or not CERT_HOST_PATTERN.match(parsed.hostname or ""):
            raise create_webhook_error(
                "SENDFN_WEBHOOK_MESSAGE_INVALID",
                "SNS message is malformed",
            )

    def _default_confirm_subscription(self, url: str) -> None:
        try:
            with urlopen(url, timeout=5) as response:
                response.read()
        except Exception as exc:  # pragma: no cover - network failure
            raise create_webhook_error(
                "SENDFN_WEBHOOK_CONFIRMATION_FAILED",
                "SNS subscription confirmation failed",
            ) from exc

    def _default_fetch_certificate(self, url: str) -> str:
        try:
            with urlopen(url, timeout=5) as response:
                payload = response.read().decode("utf-8")
        except Exception as exc:  # pragma: no cover - network failure
            raise create_webhook_error(
                "SENDFN_WEBHOOK_SIGNATURE_INVALID",
                "SNS signature verification failed",
            ) from exc
        return str(payload)

    def _default_verify_signature(
        self, canonical_message: str, signature: str, certificate: str
    ) -> bool:
        try:
            signature_bytes = base64.b64decode(signature, validate=True)
            parsed_certificate = x509.load_pem_x509_certificate(certificate.encode("utf-8"))
            public_key = parsed_certificate.public_key()
            if not isinstance(public_key, rsa.RSAPublicKey):
                return False
            public_key.verify(
                signature_bytes,
                canonical_message.encode("utf-8"),
                padding.PKCS1v15(),
                hashes.SHA1(),
            )
            return True
        except (InvalidSignature, TypeError, ValueError):
            return False
